### PR TITLE
[FLINK-19035] Remove fold from DataStream API 

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Removes deprecated DataStream#fold and all related classes.

This was voted in https://lists.apache.org/thread.html/rc1cd927939fcdce3a2dca9f509b48be0740c50b73304c2d46c6a4347%40%3Cdev.flink.apache.org%3E

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no /** don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by dawidwys** [comment]
cc @aljoscha

File: flink-streaming-scala/pom.xml [hunk 1]
```diff
@@ -262,6 +262,22 @@ under the License.
...
+							<exclude>org.apache.flink.streaming.api.scala.WindowedStream#apply(java.lang.Object,org.apache.flink.api.common.functions.FoldFunction,org.apache.flink.streaming.api.scala.function.WindowFunction,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+							<exclude>org.apache.flink.streaming.api.scala.WindowedStream#apply(java.lang.Object,scala.Function2,scala.Function4,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+							<exclude>org.apache.flink.streaming.api.scala.WindowedStream#fold(java.lang.Object,scala.Function2,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+							<exclude>org.apache.flink.streaming.api.scala.WindowedStream#fold(java.lang.Object,org.apache.flink.api.common.functions.FoldFunction,org.apache.flink.streaming.api.scala.function.WindowFunction,org.apache.flink.api.common.typeinfo.TypeInformation,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+							<exclude>org.apache.flink.streaming.api.scala.WindowedStream#fold(java.lang.Object,scala.Function2,scala.Function4,org.apache.flink.api.common.typeinfo.TypeInformation,org.apache.flink.api.common.typeinfo.TypeInformation)</exclude>
+
```
Comment (twalthr) [inline flink-streaming-scala/pom.xml:278]: nit: remove empty line and add a comment to this section that fold was dropped in 1.12

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AllWindowTranslationTest.scala [hunk 2]
```diff
@@ -91,54 +91,10 @@ class AllWindowTranslationTest {
...
-    try
-      windowedStream.fold("", new FoldFunction[String, String]() {
-        @throws[Exception]
-        def fold(accumulator: String, value: String): String = accumulator
-      })
-
-    catch {
-      case _: UnsupportedOperationException =>
-        // expected
-        // use a catch to ensure that the exception is thrown by the fold
```
Comment (twalthr) [inline flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AllWindowTranslationTest.scala:135]: this comment still occurs in the file

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala [hunk 3]
```diff
@@ -35,10 +35,13 @@ import org.apache.flink.streaming.runtime.partitioner._
...
 import org.apache.flink.util.Collector
 
+import org.hamcrest.CoreMatchers.equalTo
 import org.junit.Assert._
 import org.junit.rules.ExpectedException
 import org.junit.{Rule, Test}
 
+import java.lang
+
 class DataStreamTest extends AbstractTestBase {
```
Comment (twalthr) [inline flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala:45]: update `testResource()` in this class

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/WindowTranslationTest.scala [hunk 4]
```diff
@@ -92,56 +93,10 @@ class WindowTranslationTest {
...
-    try
-      windowedStream.fold("", new FoldFunction[String, String]() {
-        @throws[Exception]
-        def fold(accumulator: String, value: String): String = accumulator
-      })
-
-    catch {
-      case _: UnsupportedOperationException =>
-        // expected
-        // use a catch to ensure that the exception is thrown by the fold
```
Comment (twalthr) [inline flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/WindowTranslationTest.scala:138]: this comment is does still exist in the class

**Review by twalthr — APPROVED** [review]
Thanks for the PR @dawidwys. I found only minor things. Good to merge otherwise.

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/WindowTranslationTest.scala (hunk 4 above)
Comment (twalthr) [reply flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/WindowTranslationTest.scala:138]: same for java

File: flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AllWindowTranslationTest.scala (hunk 2 above)
Comment (twalthr) [reply flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/AllWindowTranslationTest.scala:135]: same for java

**Review by gaoyunhaii — APPROVED** [review]
Very thanks for @dawidwys for the change and it is indeed not easy to find all the occurrences. It is also LGTM.

File: docs/dev/stream/state/queryable_state.md [hunk 5]
```diff
@@ -155,7 +150,7 @@ descriptor.setQueryable("query-name"); // queryable state name
 </div>
 
 This variant has no limitations as to which type of state can be made queryable. This means that this can be used for 
-any `ValueState`, `ReduceState`, `ListState`, `MapState`, `AggregatingState`, and the currently deprecated `FoldingState`.
+any `ValueState`, `ReduceState`, `ListState`, `MapState`, and `AggregatingState`.
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply docs/dev/stream/state/queryable_state.md:153]: I would appreciate help with updating the Chinese docs once I merge this PR.

**Pull request merged**